### ISSUE TITLE
a problem with m.route

### ISSUE THREAD

**Issue by nordfjord** [lhorie/mithril.js#451]
Currently if I were to use m.route as such:

```js
m.route(document, '/', {
    '/:variable/:route': someModule,
    '/non/variable': anotherModule
});
```

then trying to access `/?/non/variable` would direct me to `someModule` when the expected behaviour would be directing me to `anotherModule`

This is so far from expected that I spent an hour debugging what I might have done wrong.

I'm not sure how you like your pull requests so I'm just going to add a little code example for a solution

Here I basically sort the routes by how many variables they have, to make sure those with the most variables always end up last in the route object. I'm not sure if this is the correct way to do it a better way might be to check if an exact match is found in the router object when calling routeByValue, but doing it in m.route is a one time performance hit.

```js
//inside m.route
else if (arguments.length === 3 && type.call(arguments[1]) === STRING) {
    var root = arguments[0], defaultRoute = arguments[1], oldrouter = arguments[2];
    var router = {};
    Object.keys(oldrouter).sort(function(a, b){
        return (a.match(/:/g) || []).length - (b.match(/:/g) || []).length;
    }).forEach(function(key){
        router[key] = oldrouter[key];
    });
    redirect = function(source) {
        var path = currentRoute = normalizeRoute(source);
        if (!routeByValue(root, router, path)) {
            m.route(defaultRoute, true)
        }
    };
    var listener = m.route.mode === "hash" ? "onhashchange" : "onpopstate";
    window[listener] = function() {
        if (currentRoute != normalizeRoute($location[m.route.mode])) {
            redirect($location[m.route.mode])
        }
    };
    computePostRedrawHook = setScroll;
    window[listener]()
}
```

Regards,
Einar

**Comment by ginus** [comment]
Route be matched follow the list you defined, so  just switch two line defined as 
```
m.route(document, '/', { 
   '/non/variable': anotherModule  //strict first
   '/non/:variable/': someModule,
   '/:variable/:route': someModule, 
});
```
Be care of the orders.

**Comment by nordfjord** [comment]
I know it can be solved that way, but I still see this as a problem, for a new user this isn't apparent, and can lead to frustration.

Imagine another scenario, you are adding routes to your application based on user authentication. Some of them variable routes some not. Something like this:

```js
var routes = {
    '/': mainModule,
    '/non/variable': anotherModule
    '/non/:variable/': someModule,
    '/:variable/:route': someModule
};

if( currentUser.credentials === 'ADMIN' ){
    routes['/administration/config'] = adminModule;
}

m.route(document, '/', routes);
```
What is the expected behaviour here? Will `/administration/config` be at the end of the `for in` loop in m.route or in some other arbitrary order?

What I mean to say is that this is a problem and it is trivial to solve, and in my view is something the framework should do, not the developer.

**Comment by barneycarroll** [comment]
The real issue is that the user has no control over the order of the loop. You'd assume the paths to be tested in the order they're specified, but according to ECMAscript specification, there is no mandated order for how to iterate through objects. ES6 gives us Maps, which can be emulated with an Array structure (in fact, this is how Map constructor syntax is expressed):
```Javascript 
[ [ key1, val1 ], [ key2, val2 ] ]
```
Slightly more verbose, but it gives total control of test order to the user while keeping core logic unopinionated and simple.

**Comment by nordfjord** [comment]
Then for now let's not solve it with sorting an object, let's do it in routeByValue, you can see a proposal in my pull request #452

**Comment by lhorie** [comment]
merged #452

This change will be released in the next version

**Comment by barneycarroll** [comment]
Did this ever make it in to master @lhorie?
